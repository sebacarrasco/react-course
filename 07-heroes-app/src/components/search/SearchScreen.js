import React, { useMemo } from 'react'
import { useLocation } from 'react-router-dom';
import queryString from "query-string"
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    // Este hook es para obtener la ubicación
    // En la propiedad search viene un string con los parámetros de búsqueda
    // y con queryString se parsea para tener un objeto
    const { search } = useLocation();
    const { q = "" } = queryString.parse(search); // Se le define un valor por defecto al q

    const [ { searchText }, handleInputChange ] = useForm({ searchText: q });
    // Este usememo es para que solo se haga la búsqueda una vez se hace submit
    // y cada vez que cambie el input
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    }

    return (
        <div>
            <h1>Search a hero</h1>
            <hr/>

            <div className="row">

                <div className="col-5">
                <h4>Search Form</h4>
                <hr/>
                <form onSubmit={ handleSearch }>
                    <input
                        placeholder="Find your hero"
                        className="form-control"
                        autoComplete="off"
                        onChange={ handleInputChange }
                        name="searchText"
                        value={ searchText }
                    />

                    <button
                        type="submit"
                        className="btn m-1 btn-block btn-outline-primary"
                        onSubmit={ handleSearch }
                    >
                        Search
                    </button>
                </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>

                    {
                        (q === "")
                        && <div className="alert alert-info">
                            Search a hero
                        </div>
                    }

                    {
                        ( q !== "" && heroesFiltered.length === 0 )
                        && <div className="alert alert-danger">
                            {`There is no hero with ${q}`}
                        </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}  
                            />
                        ))
                    }
                </div>

            </div>

        </div>
    )
}
