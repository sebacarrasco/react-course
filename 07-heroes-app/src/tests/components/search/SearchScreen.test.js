import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';


describe('Tests for <SearchScreen />', () => {
    
    const history = {
        push: jest.fn()
    }


    test('should show correctly with default values', () => {
       
        const wrapper = mount(
            // Como se trabaja con rutas en este componente, es necesario
            // ponerlo dentro de un router y en una ruta
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={ props => <SearchScreen history={ history }/> }/>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Search a hero");

    });

    test('should show Batman and the input with the value of the queryString', () => {
       
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route path="/search" component={ props => <SearchScreen history={ history }/> }/>
            </MemoryRouter>
        );

        expect(wrapper.find("input").prop("value")).toBe("batman");
        expect(wrapper).toMatchSnapshot();

        
    });

    test('should show an error if a hero is not found', () => {
       
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <Route path="/search" component={ props => <SearchScreen history={ history }/> }/>
            </MemoryRouter>
        );

        expect(wrapper.find(".alert-danger").text().trim()).toBe(`There is no hero with batman123`);
        expect(wrapper).toMatchSnapshot();
        
    });

    test('should call history.push', () => {
       
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <Route path="/search" component={ props => <SearchScreen history={ history }/> }/>
            </MemoryRouter>
        );

        wrapper.find("input").simulate("change", {
            target: {
                name: "searchText",
                value: "bat"
            }
        });

        wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
        
        expect(history.push).toHaveBeenCalledWith("?q=bat");
    });
    
    


});

