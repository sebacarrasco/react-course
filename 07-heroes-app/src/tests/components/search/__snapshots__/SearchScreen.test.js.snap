// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <SearchScreen /> should show Batman and the input with the value of the queryString 1`] = `
<div>
  <h1>
    Search a hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          onSubmit={[Function]}
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                 
                Batman
                 
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                More...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests for <SearchScreen /> should show an error if a hero is not found 1`] = `
<div>
  <h1>
    Search a hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          value="batman123"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          onSubmit={[Function]}
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no hero with batman123
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests for <SearchScreen /> should show correctly with default values 1`] = `
<div>
  <h1>
    Search a hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          onSubmit={[Function]}
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
